Sample object

rosborn opened the pull request:
Sample object
- Generalizes handling a sample from a model
- Read and write to file

Model object
- Creates a distribution over a set of parameters

statistics function
- Generate statistics for a sample and optionally store them in the sample

Example:
_Create a sample of size 10 over the Erdos-Renyi distribution on monomials over the ring Q[x,y] with D=4, and p=0.5._
`sample(ER(QQ[x,y],4,0.5),10)`
_Create a sample of size 15 over the Graded Erdos-Renyi distribution on monomials over the ring C[z_1..z_8] with D=2, and p={0.25,0.5}._
`sample(ER(CC[z_1..z_8],2,{0.25,0.75}),15)`
_Calculate the mean, std dev, and tally of the dimension of the ideals generated by the sample_
`statistics(sample(ER(CC[z_1..z_8],5,0.1),100))`

TODO:

- [ ] Finish documentation
- [x] Test statistics function

rosborn committed:
Added function to read Sample from file

rosborn committed:
Merge branch 'master' into sample-object

* master: (48 commits)
  Minor changes to documentation
  Update to randomMonomialIdeals examples
  Update RandomMonomialIdeals.m2
  Create RandomMonomialIdeals.m2
  Update to randomMonomialIdeals documentation
  Updated documentation to randomMonomialIdeals
  potentially fixed
  updated method regStats in RMIs.m2 to prepare for transfer to package
  Update RandomMonomialIdeals.m2
  updated to allow graded fixed M model
  Create RandomMonomialIdeals.m2
  fixed documentation as per discussion.
  2 typos fixed
  edited List description using Despina's text
  minor typo fix
  check it out: we don't need a name for the list!!
  sorry I pushed a typo! fixed.
  here are my proposed changes for documentation for pOrM
  minor language change
  minor changes
  ...

rosborn committed:
Added optional argument and additional outputs for statistics

rosborn committed:
Generalized sample to generic models

rosborn committed:
Merge branch 'master' into sample-object

* master: (43 commits)
  added doc lines for PolynomialRing input
  Changed ideals to monomials
  Add overloads for PolynomialRing
  Eliminate extra blank lines
  make it look nice
  final edits to tests
  added to dim example
  added to example
  fixed examples
  Revert "Changed 1 to 2 in Tests"
  Changed 1 to 2 in Tests
  removed 2 junk files
  added tests to check if ideals are in same ring
  standard dev in documentation
  put sets into same ring in randomMonomialSets function
  added standard deviation to dimStats and degStats
  comment on tests.
  no need to sum a sequence by hand; M2 can do it.
  matching change in dimStats documentation (for output explanation)
  minor documentation changes
  ...

rosborn committed:
Created overloads for ER and began documentation

rosborn committed:
Merge branch 'master' into sample-object

* master: (121 commits)
  commented out the unnecessary from file
  removed garbage
  function allLaurentMonomials with given 1-norm
  new file w/ example
  Merge branch 'master' into task26
  Revert "merge fixes"
  merge fixes
  Merge branch 'master' into Task-24
  Deletion of example
  small edit
  Changed so a,b,c instead of x1,x2,x3
  added another example
  small doc edits
  removed ";" from doc example
  Wording + fixed print position for showtally
  let the user see the ideals in doc
  changed a doc example
  fixed running error, added stdDev to tests
  I found the "ideals" error.
  added stdDev
  ...

rosborn committed:
Fix std deviation formula

sonjapetrovic committed:
removed 1 redundant "hash table" in doc.

rosborn committed:
Add error checks, eliminate unused options

rosborn committed:
Indent long line

rosborn committed:
Add test for statistics

rosborn committed:
Fix tests

rosborn committed:
Setup documentation

rosborn committed:
Add documentation

rosborn committed:
Add error checks

rosborn committed:
Clean up documentation

rosborn committed:
Begin documentation for ER model

rosborn committed:
Merge branch 'master' into sample-object

* master:
  change verbose message for regStats
  Merge branch 'master' into TASK30-consistency
  master merge clean-up, part 2.
  loading fixes
  Changed listOfIdeals to ideals in functions and Documentation
  Fixed Typos
  improve doc
  fixed compile error. (oops)
  making Verbose print stmts a bit more informative.
  added spacing
  fixed mingenStats verbose message
  put together verbose method for mingenStats
  made small change
  trying something else
  added space in message
  trying something
  fixed stdio for idealsFromGeneratingSets
  added verbose to regStats
  added documentation for Verbose
  added verbose to all methods

rosborn committed:
Merge branch 'master' into sample-object

sdgpfo commented:
Thanks Richard. @rosborn  I've checked the methods written, everything seems okay. 
I would suggest that ER changed to rmi or something to refer to the fact that these are random monomial ideals, since Model object is meant to be general, and Erdos-Renyi type models can be defined (and have been defined) on many different objects.)

Also there is only one test written, for the statistics function: is there currently an intention to write additional tests, with pre-generated sample objects for example and to test the other methods also?

sonjapetrovic committed:
to do: fix terminology slightly

sonjapetrovic committed:
Merge branch 'master' into sample-object

sonjapetrovic committed:
seems that there was a ; missing in an odd method.

Sondzus commented:
Here's an example I would like to run (and include in the JSAG paper ?? ) @dwilburne @sdgpfo - is this overkill or should we try to include an example like this? 
[ * update 7/17/17:
Now I think it may be overkill. We can showcase the use of statistics without the use of a new Model. We are setting up a framework within which people can define models formally; so why would we want to include an example of a "silly model" which is not really formal? I am going to make edits in the paper today to reflect this. Still, for the future, the comments below hold.
* ] 

**the following should work IF Model is exported; it is currently not:
-- make a new model: 
sillyModel=method(TypicalValue=>Model)
sillyModel(ZZ,ZZ):=(n,D)->(
    x := symbol x;
    R := QQ[x_1..x_n];
    tbl := new MutableHashTable;
    tbl.Name = "random polynomials of degree D in n variables";
    tbl.Parameters = (n,D);
    tbl.Generate = ()->random(D,R);
    new Model from tbl
)
--sample from it:
mySample = sample(oo,10)
--get statisitcs on num of gb elements of my sample:
statistics(mySample, numcols@@ gens@@gb@@ideal)

BUT i get:
i103 : mySample = sample(myModel,10)
stdio:178:12:(3):[1]: error: new: expected a hash table as prospective class

sonjapetrovic committed:
should we export Model and Sample?

sonjapetrovic committed:
a question.

sonjapetrovic committed:
attempt

had a baby so i’m just uploading the draft of what i tried to finish;
it’s not finished, sorry.

sonjapetrovic committed:
standardized some documentation

.. and noticed a lot more TO-DOs in there. still left to do…

Sondzus commented:
@dwilburne @sdgpfo  I am looking back at the last things I did and the few questions I left. One of them was whether we should export model and sample, and to that end I made an attempt of an example to make a new model. I got an error and documented it. I suppose @rosborn  could help but I don't know if he has the time. 
So we may just leave this as is and not include that attempted example. (thought I liked it and ultimately I do want to know how to do it anyway!)

In other news, there are several TO DO items in the documentation for the sample object; they are very small and probably require a couple of sentences. If you two don't get to them I will do my best to finish them off next week.

sonjapetrovic committed:
not exporting model(). also fixed but in statistics.

we don’t have a good example for exporting model generating method.
(also i have issues doing it in m2 with some hashtable errors.)

one test was failing so i fixed statistics().

sonjapetrovic committed:
doc missing for Model and Sample types

even if we don’t make a method that creates a new Model, shouldn’t we
export Model and Sample? and document them as new types.
except i don’t know how to do this properly…

sonjapetrovic committed:
added doc for Model and Sample.

very basic. this should be enough.

dwilburne committed:
ER model type doc

Added documentation for ER model types

dwilburne committed:
sample/model doc edits

minor edits to model doc nodes

dwilburne committed:
sample object doc update

edited examples in sample object doc and added some SeeAlso's

dwilburne committed:
fixed typo in sample doc

typo fix
